### PR TITLE
fix(memory): address PR #139 review feedback

### PR DESCRIPTION
**Pull request by jesseturner21**
## Description

Addresses all review comments from PR #139. Key fixes:

- **`getLastKTurns` correctness** — was returning the first K turns instead of the last K. Now paginates all events and slices the tail.
- **`findMemoryIdByName` collision** — prefix match `prod` could match `prod-east-xxx`. Tightened to require a single alphanumeric suffix segment.
- **`waitForMemories` error handling** — was swallowing all errors (including auth/config). Now only swallows transient errors (throttling, 5xx).
- **`pollUntil` edge cases** — `maxWaitSeconds: 0` now runs at least one attempt; no longer sleeps past deadline.
- **`credentialsProvider` → `credentials`** — renamed to match AWS SDK v3 convention.
- **`maxAttempts`/`retryMode`** — added to `MemoryClientConfig` and passed through to SDK clients.
- **`./memory` export** — added to `package.json` exports map.
- **JSDoc** — added `@returns` documentation to `waitForMemories`.

## Related Issues

Addresses review comments on #139

## Type of Change

Bug fix

## Testing

How have you tested the change?

- [x] I ran `npm run check`
- [x] Added unit tests for last-K-turns ordering (10 turns, request k=3, assert turns 8-10)
- [x] Added unit test for name-collision rejection (prod vs prod-east)
- [x] Added unit tests for pollUntil edge cases (maxWaitSeconds=0, no sleep past deadline)

## Checklist
- [x] I have read the CONTRIBUTING document
- [x] I have added any necessary tests that prove my fix is effective or my feature works
- [x] I have updated the documentation accordingly
- [x] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/memory/client.ts [hunk 1]
```diff
@@ -39,7 +39,9 @@ class MemoryPassthroughClient {
     const region = config?.region ?? process.env.AWS_REGION ?? DEFAULT_REGION
     const clientConfig = {
       region,
-      ...(config?.credentialsProvider && { credentials: config.credentialsProvider }),
+      ...(config?.credentials && { credentials: config.credentials }),
```
Comment (jesseturner21) [inline src/memory/client.ts:42]: **Fixes review comment #7 (rename `credentialsProvider` → `credentials`) and #6 (add retry options).**

Renamed to match the AWS SDK v3 convention where the field is always `credentials`. Also passes `maxAttempts` and `retryMode` through to the underlying SDK clients so customers can configure retry behavior for batch-flush scenarios.

File: src/memory/client.ts [hunk 2]
```diff
@@ -139,7 +148,14 @@ class MemoryClient extends MemoryPassthroughClient {
       {
         maxWaitSeconds: params.maxWaitSeconds ?? 180,
         pollIntervalMs: params.pollIntervalMs ?? 15_000,
-        shouldSwallowError: () => true,
+        shouldSwallowError: (err) => {
+          const name = (err as { name?: string })?.name
+          return (
+            name === 'ThrottlingException' ||
+            name === 'ServiceUnavailableException' ||
```
Comment (jesseturner21) [inline src/memory/client.ts:155]: **Fixes review comment #4 (scope error swallowing in `waitForMemories`).**

Previously `shouldSwallowError: () => true` ate all errors including auth/config errors, causing a silent 180s timeout. Now only transient service errors (ThrottlingException, ServiceUnavailableException, InternalServerException) are retried. Config errors like AccessDeniedException or ResourceNotFoundException propagate immediately.

File: src/memory/client.ts [hunk 3]
```diff
@@ -159,39 +175,33 @@ class MemoryClient extends MemoryPassthroughClient {
       }
     }
 
+    const allEvents = await paginateAll(
```
Comment (jesseturner21) [inline src/memory/client.ts:178]: **Fixes review comment #1 (`getLastKTurns` returns first K, not last K).**

The old implementation paged forward and stopped at K, returning turns 1–K. Now fetches all events via `paginateAll`, groups into turns, and returns `slice(-k)` to get the actual last K turns. This is O(n) in session length — a service-side tail API would be needed to optimize for very long sessions.

File: src/_utils/polling.ts [hunk 4]
```diff
@@ -16,12 +16,13 @@ export async function pollUntil(
   }
 ): Promise<boolean> {
   const deadline = Date.now() + opts.maxWaitSeconds * 1000
-  while (Date.now() < deadline) {
+  while (true) {
```
Comment (jesseturner21) [inline src/_utils/polling.ts:19]: **Fixes review comment #5 (`pollUntil` edge cases).**

Two fixes: (1) `maxWaitSeconds: 0` now runs at least one attempt before checking the deadline, instead of immediately returning false. (2) The deadline check happens before sleeping, so we never waste a full `pollIntervalMs` sleep after the deadline has passed.

File: src/memory/types.ts [hunk 5]
```diff
@@ -33,7 +33,10 @@ export type ControlPlaneMethods = Pick<BedrockAgentCoreControl, (typeof CONTROL_
 
 export interface MemoryClientConfig {
   region?: string
-  credentialsProvider?: AwsCredentialIdentityProvider
+  /** Credential provider function (e.g. fromSSO, fromTemporaryCredentials). */
+  credentials?: AwsCredentialIdentityProvider
```
Comment (jesseturner21) [inline src/memory/types.ts:37]: **Fixes review comments #6 and #7.**

Added `maxAttempts` and `retryMode` for retry configuration (comment #6), and renamed `credentialsProvider` → `credentials` for AWS SDK v3 naming consistency (comment #7).

File: package.json [hunk 6]
```diff
@@ -53,6 +53,10 @@
     "./identity": {
       "import": "./dist/src/identity/index.js",
       "types": "./dist/src/identity/index.d.ts"
+    },
+    "./memory": {
+      "import": "./dist/src/memory/index.js",
+      "types": "./dist/src/memory/index.d.ts"
     }
```
Comment (jesseturner21) [inline package.json:60]: **Fixes review comment #3 (missing `./memory` export).**

Adds the `bedrock-agentcore/memory` import path so downstream consumers (CLI templates, AgentCoreMemoryPlugin) can import directly.

File: src/memory/client.ts [hunk 7]
```diff
@@ -234,7 +244,12 @@ async function findMemoryIdByName(controlPlane: BedrockAgentCoreControl, name: s
   let nextToken: string | undefined
   do {
     const resp = await controlPlane.listMemories({ nextToken })
-    const match = resp.memories?.find((m) => m.id?.startsWith(`${name}-`) || m.id === name)
+    const match = resp.memories?.find((m) => {
+      if (m.id === name) return true
+      if (!m.id?.startsWith(`${name}-`)) return false
+      const suffix = m.id.slice(name.length + 1)
+      return /^[a-z0-9]+$/i.test(suffix)
```
Comment (jesseturner21) [inline src/memory/client.ts:253]: Fixes review comment #2 (createOrGetMemory name-collision bug). The old heuristic matched any id starting with the name prefix. Now requires the suffix to be a single alphanumeric segment, preventing cross-name collisions.

**Review by aidandaly24 — APPROVED** [review]
